Add --link flag to compile to allow linking through BSP

KristianAN opened the pull request:
Before this commit linking could only be done through the run command with BSP. Because the BSP protocol does describe a linking command (as it is often a part of the compile step) the flag --link has been added to the compile command. This is useful for those kinds of projects that need the linker to be run to produce source code for further toolchains, e.g. a ScalaJS project might often use @JSExport and then run the code from some other JS entrypoint.

KristianAN commented:
@tgodzik  I'm not quite sure where/how I can write a test for this. I looked at BspCompileSpec, but I'm not quite sure how I can pass in the compileArgs there.
 
 I also want to further investigate some bugs with the JS linker config as it does not seem to respect everything that is set from bloop-config, but maybe that should be it's own PR?

KristianAN commented:
this should also fix: https://github.com/scalacenter/bloop/issues/1304

KristianAN commented:
The JS-bridge defines this:

```scala
    val moduleInitializers = mainClass match {
      case Some(mainClass) if runMain =>
        logger.debug(s"Setting up main module initializers for $project")
        List(ModuleInitializer.mainMethodWithArgs(mainClass, "main"))
      case _ =>
        if (runMain) {
          logger.debug(s"Setting up no module initializers, commonjs module detected $project")
          Nil // If run is disabled, it's a commonjs module and we link with exports
        } else {
          // There is no main class, install the test module initializers
          logger.debug(s"Setting up test module initializers for $project")
          List(
            ModuleInitializer.mainMethod(
              TestAdapterInitializer.ModuleClassName,
              TestAdapterInitializer.MainMethodName
            )
          )
        }
    }
```
I disagree here. Not running main is indicative of JSExport and the likes I think, not that we want to link with tests (I don't think this is very common). I propose here that we instead add a test flag for the test case and add that same flag to the Commands.Test (it can also be an additional flag in the compileArgs).

KristianAN commented on frontend/src/main/scala/bloop/bsp/BloopBspServices.scala:
```diff
@@ -421,6 +424,48 @@ final class BloopBspServices(
+          (state, result)
+        }
+      case Jvm(_, _, _, _, _, _) =>
+        compileProjects(userProjects :: Nil, state, compileArgs, originId, logger)
```
Alternatively here we can do the same as in the interpreter when it hits a JVM project on link. I'm not sure if there is much difference.

tgodzik commented:
> @tgodzik I'm not quite sure where/how I can write a test for this. I looked at BspCompileSpec, but I'm not quite sure how I can pass in the compileArgs there.

This should be similar to https://github.com/scalacenter/bloop/commit/32dbfb14dbaec488eccb515db67a2b0df8339f54#diff-9458dec34378d1535538ca59de86e441bcef70b5739a0cc484281ac8ed1ef1aaR695

> I also want to further investigate some bugs with the JS linker config as it does not seem to respect everything that is set from bloop-config, but maybe that should be it's own PR?

Would probably be easier to review, but I can also ask for help reviewing

tgodzik commented:
> I disagree here. Not running main is indicative of JSExport and the likes I think, not that we want to link with tests (I don't think this is very common). I propose here that we instead add a test flag for the test case and add that same flag to the Commands.Test (it can also be an additional flag in the compileArgs).

I haven't worked in that part too much, but I think that might have been a simplification since Bloop wasn't used for publishing anything

KristianAN commented on frontend/src/main/scala/bloop/bsp/BloopBspServices.scala:
```diff
@@ -421,6 +424,84 @@ final class BloopBspServices(
+          case Success(linkRunsSeq) =>
+            linkRunsSeq
+              .map(_._2)
+              .foldLeft(Option.empty[CompileResult])((_, res) => // TODO propogate stuff better here
```
This logic is somewhat lifted from the test function, but it doesn't really work that well here. I'm not entirely sure what the best strategy is for combining the compileresults (actually linker results)

KristianAN commented:
So far I have tested that the changes in this PR allows for linking using JSExport and with a main method without having to call run.

This is also tested through BSP in relation to this PR: https://github.com/oyvindberg/bleep/pull/445.

I have also added support for the new minification in ScalaJS which is now set to true for release esmodule builds. We can add an option for this in bloop-config so that users can override this setting if they want.

Further TODOs for this PR:

- Fix the TODO I have commented
- Add some tests
- Look at the ScalaNative linker

@tgodzik anything else you think should be done?

tgodzik commented:
> anything else you think should be done?

Nothing comes to mind, but maybe @sjrd will be able to take a quick look here?

sjrd commented on bridges/scalajs-1/src/main/scala/bloop/scalajs/JsBridge.scala:
```diff
@@ -85,7 +85,7 @@ object JsBridge {
         case (ModuleKindJS.ESModule, Some(value)) =>
           StandardImpl.clearableLinker(
-            linkerConfig.withModuleSplitStyle(value)
+            linkerConfig.withModuleSplitStyle(value).withMinify(isFullLinkJS)
```
In order to be consistent with sbt-scalajs, you should enable `minify` for *all* `fullLink` builds, regardless of `ModuleKind`.

Even when used in conjunction with the Closure compiler, `minify` still brings some code size improvements: the optimizer works extra hard to reduce the need for temporary local variables.

KristianAN commented on frontend/src/main/scala/bloop/bsp/BloopBspServices.scala:
```diff
@@ -421,6 +424,76 @@ final class BloopBspServices(
+            }
+            (linkState, result)
+          }
+        case Jvm(_, _, _, _, _, _) => ???
```
@tgodzik Should we error here or just do a NoOp? if we do a NoOp a client can request linking all projects and only the ones that can link will be linked and trying to link JVM doesn't stop the linking. If we error  the client has to be explicit and only try to link targets that can be linked.

tgodzik replied:
We should probably just compile normally in this case and ignore link option. A warning could be printed

tgodzik commented:
> @tgodzik this is now blocked by [scalacenter/bloop-config#102](https://github.com/scalacenter/bloop-config/pull/102)

Running the release here https://github.com/scalacenter/bloop-config/actions/runs/12394117349

KristianAN commented on bridges/scalajs-1-test/src/test/scala/bloop/scalajs/ScalaJsToolchainSpec.scala:
Moving the these tests to a different module lets us remove the frontend dependency from the bridges. This allows us to create working tests for bsp in the bridges. However it breaks the ToolchainSpecs because the loading the bridge toolchain with the classloader no longer works. 

I am no guru on these classloader issues.  This is the part of the test that no longer works when separating the modules.

```scala
  val state0: State = {
    def setUpScalajs(p: Project): Project = {
      val platform = p.platform match {
        case jsPlatform: Platform.Js =>
          jsPlatform.copy(toolchain = Some(ScalaJsToolchain.apply(this.getClass.getClassLoader)))
        case _ => p.platform
      }
      p.copy(platform = platform)
    }

    val configDir = TestUtil.getBloopConfigDir("cross-test-build-scalajs-1.x")
    val logger = bloop.logging.BloopLogger.default(configDir.toString())
    TestUtil.loadTestProject(configDir, logger, true, _.map(setUpScalajs))
  }
```
@tgodzik do you have any neat trick to make this work?

KristianAN replied:
Actually I think the solution here is just to move these tests into frontend

KristianAN commented:
Chages introduced in PR:
  - new --link flag for bsp compile, which tries to link the given projects, compiles jvm projects like normal if a jvm project is supplied
  - Removed dependency on frontend in all bridges
  - Moved CrossPlatform.scala to shared
  - Moved all bridge tests to frontend
  - frontend local publishes all bridges before running tests
  - New tests for linking projects through bsp
  - New version of bloop-config that allows clients to specify which type of native project is being linked (application, dynamiclibrary, staticlibrary)
  - Changes made to linking logic that now allows for linking without specifying a main class. This is mainly useful for ScalaJS projects with JSExport and for scala-native projects using dynamicLibrary or staticLibrary application type.
  
Overall This should bring bloop much closer to sbt/mill and the likes with regards to linking capabilities. There is probably still more work we can do to better the client capabilities when it comes to passing configuration, but I think we can add that later. Hopefully this PR will let us see more linking through bloop so we can discover possible bugs etc.

KristianAN commented:
One thing I forgot to add is that I have not really done much with the scalajs bridge for 0.6. I have not added tests, or decoupled it from frontend like the others. I highly doubt there are many projects using bloop for such an old version. It should work as it did before however.

tgodzik commented on build.sbt:
```diff
@@ -366,7 +369,7 @@ lazy val jsBridge06 = project
 val jsBridge1Name = "bloop-js-bridge-1"
 lazy val jsBridge1 = project
-  .dependsOn(frontend % Provided, frontend % "test->test")
+  .dependsOn(bloopShared % Provided /* frontend % "test->test" */ )
```
if no longer needed we can remove the test dependency

tgodzik commented on frontend/src/main/scala/bloop/bsp/BloopBspServices.scala:
```diff
@@ -421,6 +424,90 @@ final class BloopBspServices(
+    val isDebug = compileArgs.exists(_ == "--debug")
+
+    val overrideLinkerMode = if (isRelease && isDebug) {
+      None
```
MAybe we should log an error in this case?

KristianAN replied:
I logged it as a warning as we still run the linking, we just ignore the dynamic --debug or --release arguments.

tgodzik reviewed (changes requested):
Looks good! Just two minor comments, otherwise LGTM

tgodzik reviewed (approved):
Great work! LGTM
